lib/ansible/modules/system/etc_hosts.py

```diff
@@ -0,0 +1,342 @@
+    def __init__(self, module):
+        # To be able to use fail_json
+        self.module = module
+
+        # Shortcut for params
+        self.params = module.params
+
+        # This is where we will store all the hosts file lines
+        self.lines = []
+
+        self.diff = {
+            'before': '',
+            'after': ''
+        }
+
+        # Read the file
+        if os.path.isfile(self.params['path']):
+            self._read()
+
+    def _read(self):
+        # Open the hosts file
+        try:
+            f = open(self.params['path'])
+        except IOError:
+            e = get_exception()
+            self.module.fail_json(
+                msg=(
+                    "Cannot open hosts file %s for reading." %
+                    self.params['path']),
+                details=str(e))
+
+        # Splitting pattern
+        p_comment = re.compile("\s*#.*")
+        p_split = re.compile("\s+")
+
+        for line in f:
+            self.diff['before'] += line
+
+            record = {
+                'valid': False,
+                'line': line
+            }
+
+            # Remove all comments from the line
+            line = p_comment.sub('', line)
+
+            if len(line) > 0:
+                # Split the line
+                items = p_split.split(line.strip())
+
+                # Check if we have valid line
+                if len(items) > 1:
+                    record['valid'] = True
+                    record['ip'] = items[0]
+                    record['hostname'] = items[1]
+
+                    # Optional alias(es)
+                    if len(items) == 3:
+                        record['alias'] = [items[2], ]
+                    elif len(items) > 3:
+                        record['alias'] = sorted(items[2:])
+
+            self.lines.append(record)
+
+        # Close the file
+        try:
+            f.close()
+        except IOError:
+            e = get_exception()
+            self.module.fail_json(
+                msg=(
+                    "Cannot close the hosts file %s after reading." %
+                    self.module['path']),
+                detail=str(e))
+
+    def add(self):
+        return self._check('add'), self.diff
+
+    def remove(self):
+        return self._check('remove'), self.diff
+
+    def _format_line(self):
+        line = "%s  %s" % (self.params['ip'], self.params['hostname'])
+
+        if self.params['alias'] is not None:
+            line += "  %s" % ' '.join(self.params['alias'])
+
+        line += "\n"
+
+        return line
+
+    def _check(self, action):
+        changed = False
+        found = None
+
+        # Check if the same record already is in the file
+        for idx, line in enumerate(self.lines):
+            if (
+                    line['valid'] and
+                    line['ip'] == self.params['ip'] and (
+                        self.params['hostname'] is None or
+                        line['hostname'] == self.params['hostname'])):
```

Comment:
I would be tempted to extract this check into a method or two. (To simplify reading the logic, and also to make it easier to easier to unit test).